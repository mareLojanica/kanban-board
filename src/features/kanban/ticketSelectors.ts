import { createSelector } from "@reduxjs/toolkit"
import { RootState } from "../../app/store"

export const selectTicketsState = (state: RootState) => state.tickets

export const selectIsLoading = createSelector(
	selectTicketsState,
	(ticketsState) => ticketsState.isLoading
)

export const selectIsError = createSelector(
	selectTicketsState,
	(ticketsState) => !!ticketsState.isError
)

export const selectSearchText = (state: RootState) => state.tickets.searchText
export const selectTickets = (state: RootState) => state.tickets.tickets
