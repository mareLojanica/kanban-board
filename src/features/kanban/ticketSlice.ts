import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import { ticketsApi } from "../../api/tickets.grapql"

interface TicketsState {
	tickets: Array<{
		id: string
		status: string
		title: string
	}>
	isLoading: boolean
	isError: boolean
	searchText: string
}

const initialState: TicketsState = {
	isLoading: false,
	tickets: [],
	isError: false,
	searchText: "",
}

const ticketsSlice = createSlice({
	name: "tickets",
	initialState,
	reducers: {
		setTickets(
			state,
			action: PayloadAction<
				{
					id: string
					status: string
					title: string
				}[]
			>
		) {
			state.tickets = action.payload
		},
		setSearchText(state, action: PayloadAction<string>) {
			state.searchText = action.payload
		},
		deleteSelectedTicket(state, action: PayloadAction<string>) {
			state.tickets = state.tickets.filter(
				(ticket) => ticket.id !== action.payload
			)
		},
		addNewTicket(
			state,
			action: PayloadAction<{
				id: string
				status: string
				title: string
			}>
		) {
			state.tickets = [...state.tickets, action.payload]
		},
		updateTicketById(
			state,
			action: PayloadAction<{ id: string; title: string }>
		) {
			state.tickets = state.tickets.map((ticket) => {
				if (ticket.id === action.payload.id) {
					return { ...ticket, title: action.payload.title }
				}
				return { ...ticket }
			})
		},
	},
	extraReducers: (builder) => {
		builder.addMatcher(
			ticketsApi.endpoints.getTickets.matchFulfilled,
			(state, { payload }) => {
				state.tickets = payload
			}
		)
	},
})

export const {
	setTickets,
	setSearchText,
	deleteSelectedTicket,
	addNewTicket,
	updateTicketById,
} = ticketsSlice.actions
export default ticketsSlice.reducer
