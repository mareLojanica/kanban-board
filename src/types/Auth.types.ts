export interface AuthState {
	accessToken: string | null
	isLoading: boolean
	error: string | null
}
