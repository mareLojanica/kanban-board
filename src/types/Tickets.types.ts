export interface Ticket {
	id: string
	status: string
	title: string
	isNew?: boolean
}

export interface TicketsState {
	tickets: Ticket[]
	isLoading: boolean
	isError: boolean
	searchText: string
}
