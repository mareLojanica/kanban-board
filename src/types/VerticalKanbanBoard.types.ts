export enum TicketStatus {
	TO_DO = "TO_DO",
	IN_PROGRESS = "IN_PROGRESS",
	DONE = "DONE",
}
export interface KanbanCardProps {
	id: string
	title: string
	index: number
	parent: string
	isNew?: boolean
	cardColor: string
}
export interface ColumnConfig {
	alias: TicketStatus
	headerColor: string
	backgroundColor: string
	cardColor: string
	title: string
}

export interface KanbanLaneProps {
	column: ColumnConfig
	items: {
		title: string
		id: string
		isNew?: boolean
	}[]
}
