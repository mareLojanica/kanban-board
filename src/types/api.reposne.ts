import { Ticket } from "./Tickets.types"
import { TicketStatus } from "./VerticalKanbanBoard.types"
import { MutationOptions, QueryOptions } from "@apollo/client"

export interface TicketResponse {
	id: string
	title: string
	status: TicketStatus
	__typename: string
}
export interface GetTicketsResponse {
	getTickets: TicketResponse[]
}

export interface UpdateTicketResponse {
	updateTicket: Omit<TicketResponse, "__typename">
}

export interface DeleteTicketResponse {
	deleteTicket: {
		success: boolean
	}
}

export interface CreateTicketResponse {
	createTicket: Ticket
}

export interface BaseQueryArgs {
	body:
		| MutationOptions<
				MutationOptions,
				Record<string, any>,
				{ headers: { Authorization: string } }
		  >
		| QueryOptions<any, Record<string, any>>
}
