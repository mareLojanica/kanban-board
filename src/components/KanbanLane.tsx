import { useDroppable } from "@dnd-kit/core"
import { Box, Typography, IconButton } from "@mui/material"
import KanbanCard from "./KanbanCard"
import AddIcon from "@mui/icons-material/Add"
import { TicketStatus } from "../types/VerticalKanbanBoard.types"
import { useCreateTicketMutation } from "../api/tickets.grapql"
import { useDispatch } from "react-redux"
import { addNewTicket } from "../features/kanban/ticketSlice"

interface ColumnConfig {
	alias: TicketStatus
	headerColor: string
	backgroundColor: string
	cardColor: string
}

interface KanbanLaneProps {
	column: ColumnConfig
	items: { title: string; id: string }[]
}

export function KanbanLane({ column, items }: KanbanLaneProps) {
	const { setNodeRef } = useDroppable({
		id: column.alias,
	})
	const dispatch = useDispatch()
	const [createTicket] = useCreateTicketMutation()

	const handleCreateTicket = async () => {
		const statusValue =
			TicketStatus[column.alias as unknown as keyof typeof TicketStatus]
		console.log(column.alias)
		const newTicket = await createTicket({
			title: "Double Click To Edit",
			status: statusValue,
		}).unwrap()
		dispatch(addNewTicket(newTicket.createTicket))
	}

	return (
		<Box
			sx={{
				flex: 3,
				display: "flex",
				flexDirection: "column",
				height: "100%",
				borderRadius: "12px",
			}}
		>
			<Box
				sx={{
					background: column.headerColor,
					color: "#ffffff",
					padding: 2,
					display: "flex",
					justifyContent: "center",
					alignItems: "center",
					borderRadius: "12px",
					width: "100%",
				}}
			>
				<Box
					sx={{
						display: "flex",
						margin: "auto",
						flexDirection: "column",
						alignItems: "center",
					}}
				>
					<Typography
						variant="h6"
						fontWeight="bold"
						sx={{ marginBottom: 1 }}
					>
						{column.alias}
					</Typography>
					<Typography
						variant="subtitle2"
						fontWeight="bold"
						sx={{ marginBottom: 1 }}
					>
						{`(${items.length})`}
					</Typography>
				</Box>
				<Box sx={{ color: "#ffffff" }}>
					<IconButton color={"inherit"} onClick={handleCreateTicket}>
						<AddIcon color={"inherit"} />
					</IconButton>
				</Box>
			</Box>

			<Box
				ref={setNodeRef}
				sx={{
					background: column.backgroundColor,
					borderRadius: 2,
					flex: 1,
					padding: 2,
					display: "flex",
					gap: 1,
					flexDirection: "column",
				}}
			>
				{items.map(({ title: cardTitle, id }, key) => (
					<KanbanCard
						id={id}
						title={cardTitle}
						key={key}
						index={key}
						parent={column.alias}
					/>
				))}
			</Box>
		</Box>
	)
}
