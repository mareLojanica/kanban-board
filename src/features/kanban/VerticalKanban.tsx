import {
	DndContext,
	DragEndEvent,
	rectIntersection,
	useSensor,
	useSensors,
} from "@dnd-kit/core"
import {
	useGetTicketsQuery,
	useUpdateTicketMutation,
} from "../../api/tickets.grapql"
import Spinner from "../../components/Spinner"
import {
	Box,
	debounce,
	Input,
	InputProps,
	styled,
	useMediaQuery,
	useTheme,
} from "@mui/material"
import { columnsConfig } from "../../config/kanbanConfig"

import { useDispatch, useSelector } from "react-redux"
import { setSearchText, setTickets } from "./ticketSlice"
import { RootState } from "../../app/store"
import { TicketStatus } from "../../types/VerticalKanbanBoard.types"
import { getEnumValueByKey } from "../../utils"
import { selectSearchText } from "./ticketSelectors"
import { useCallback, useState } from "react"
import { CustomPointerSensor } from "../../utils/CustomPointerSensor"
import KanbanLane from "../../components/KanbanLane"

const StyledInput = styled(Input)<InputProps>(({ theme }) => ({
	borderRadius: "8px",
	border: "1px solid #ccc",
	padding: "8px 12px",
	transition: "border-color 0.3s ease, box-shadow 0.3s ease",
	"&:hover": {
		borderColor: theme.palette.primary.main,
	},
	"&.Mui-focused": {
		borderColor: theme.palette.primary.main,
		boxShadow: `0 0 4px ${theme.palette.primary.main}`,
	},
	"&.Mui-error": {
		borderColor: theme.palette.error.main,
		boxShadow: `0 0 4px ${theme.palette.error.main}`,
	},
	"& input": {
		fontSize: "16px",
		color: theme.palette.text.primary,
	},
}))

const VerticalKanban = () => {
	const theme = useTheme()
	const isMobileOrTablet = useMediaQuery(theme.breakpoints.down("md"))
	const pointerSensor = useSensor(CustomPointerSensor, {
		activationConstraint: {
			delay: 100,
			tolerance: 5,
		},
	})

	const sensors = useSensors(pointerSensor)
	const dispatch = useDispatch()
	const data = useSelector((state: RootState) => state.tickets.tickets)
	const searchText = useSelector(selectSearchText)
	const [debouncedSearchText, setDebouncedSearchText] = useState(searchText)
	const { isLoading } = useGetTicketsQuery({
		searchText:
			debouncedSearchText.length > 3 ? debouncedSearchText : undefined,
	})

	const [updateTicket] = useUpdateTicketMutation()

	const handleSearchChange = useCallback(
		debounce((value: string) => {
			setDebouncedSearchText(value)
		}, 300),
		[]
	)

	if (isLoading) {
		return <Spinner />
	}

	const handleDragEnd = async (event: DragEndEvent) => {
		const draggedTo = event.over?.id
		const activeCard = event.active.id
		if (
			draggedTo &&
			event.active.data.current &&
			draggedTo !== event.active.data.current.parent
		) {
			const updatedTickets = data?.map((ticket) => {
				if (ticket.id === activeCard) {
					return { ...ticket, status: String(draggedTo) }
				} else {
					return { ...ticket }
				}
			})

			dispatch(setTickets(updatedTickets ?? []))
			const statusKey = getEnumValueByKey(TicketStatus, String(draggedTo))
			updateTicket({
				id: String(activeCard),
				status: statusKey,
			}).unwrap()
		}
	}

	const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const value = e.target.value
		dispatch(setSearchText(value))
		handleSearchChange(value)
	}

	return (
		<DndContext
			collisionDetection={rectIntersection}
			onDragEnd={handleDragEnd}
			sensors={sensors}
		>
			<Box
				sx={{
					display: "flex",
					flexDirection: "column",
					p: 2,
				}}
			>
				<Box
					sx={{
						width: "100%",
						display: "flex",
						justifyContent: "center",
					}}
				>
					<StyledInput
						placeholder="Search tickets..."
						sx={{
							width: {
								xs: "100%",
								md: "50%",
							},
						}}
						value={searchText}
						onChange={handleInputChange}
					/>
				</Box>
				<Box
					sx={{
						display: "flex",
						flexDirection: "row",
						gap: "16px",
						height: "100%",
						padding: 2,
						overflow: "auto",
					}}
				>
					{columnsConfig.map((column) => (
						<Box
							key={column.alias}
							sx={{
								width: isMobileOrTablet ? "250px" : "33%",
								minWidth: "250px",
							}}
						>
							<KanbanLane
								column={column}
								items={
									data.filter(
										(card) => card.status === column.alias
									) ?? []
								}
							/>
						</Box>
					))}
				</Box>
			</Box>
		</DndContext>
	)
}

export default VerticalKanban
