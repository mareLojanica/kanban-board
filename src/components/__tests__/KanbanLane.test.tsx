import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { Provider } from "react-redux"
import configureStore from "redux-mock-store"
import KanbanLane from "../KanbanLane"
import { TicketStatus } from "../../types/VerticalKanbanBoard.types"
import { addNewTicket } from "../../features/kanban/ticketSlice"
import { useCreateTicketMutation } from "../../api/tickets.grapql"

jest.mock("../../api/tickets.grapql", () => ({
	useCreateTicketMutation: jest.fn(),
}))

jest.mock("../KanbanCard", () => () => <div data-testid="kanban-card" />)

const mockStore = configureStore([])
const mockCreateTicket = jest.fn(() => ({
	unwrap: jest.fn().mockResolvedValue({
		createTicket: { id: "new-ticket-id", title: "Double Click To Edit" },
	}),
}))

;(useCreateTicketMutation as jest.Mock).mockReturnValue([mockCreateTicket])

const mockColumn = {
	alias: TicketStatus.TO_DO,
	title: "To Do",
	headerColor: "#3498db",
	backgroundColor: "#ecf0f1",
	cardColor: "#ffffff",
}

const mockItems = [
	{ id: "1", title: "First Task", isNew: false },
	{ id: "2", title: "Second Task", isNew: false },
]

describe("KanbanLane Component", () => {
	let store

	beforeEach(() => {
		jest.clearAllMocks()
		store = mockStore({ tickets: [] })
	})

	const renderComponent = (props = {}) =>
		render(
			<Provider store={store}>
				<KanbanLane column={mockColumn} items={mockItems} {...props} />
			</Provider>
		)

	it("renders the KanbanLane correctly", () => {
		const { container } = renderComponent()
		expect(container).toMatchSnapshot()
	})

	it("renders the column title and item count", () => {
		renderComponent()

		expect(screen.getByText("To Do")).toBeInTheDocument()
		expect(screen.getByText("(2)")).toBeInTheDocument()
	})

	it("renders the correct number of KanbanCards", () => {
		renderComponent()

		const kanbanCards = screen.getAllByTestId("kanban-card")
		expect(kanbanCards).toHaveLength(mockItems.length)
	})

	it("dispatches addNewTicket when a new ticket is created", async () => {
		renderComponent()

		const user = userEvent.setup()
		const addButton = screen.getByTestId(`add-card-${mockColumn.alias}`)

		await user.click(addButton)

		expect(mockCreateTicket).toHaveBeenCalledWith({
			title: "Double Click To Edit",
			status: TicketStatus.TO_DO,
		})

		const actions = store.getActions()
		expect(actions).toEqual(
			expect.arrayContaining([
				expect.objectContaining({
					type: addNewTicket.type,
					payload: {
						id: "new-ticket-id",
						title: "Double Click To Edit",
						isNew: true,
					},
				}),
			])
		)
	})
})
