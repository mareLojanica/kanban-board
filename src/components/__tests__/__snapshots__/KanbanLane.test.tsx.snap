// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KanbanLane Component renders the KanbanLane correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-11mey90"
    data-testid="kanban-lane-TO_DO"
  >
    <div
      class="MuiBox-root css-1vui9rc"
    >
      <div
        class="MuiBox-root css-1lf773q"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-155446w-MuiTypography-root"
        >
          To Do
        </h6>
        <h6
          class="MuiTypography-root MuiTypography-subtitle2 css-10un0gn-MuiTypography-root"
        >
          (2)
        </h6>
      </div>
      <div
        class="MuiBox-root css-g81iaw"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-cdtzx3-MuiButtonBase-root-MuiIconButton-root"
          data-testid="add-card-TO_DO"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root css-79iqgk"
    >
      <div
        data-testid="kanban-card"
      />
      <div
        data-testid="kanban-card"
      />
    </div>
  </div>
</div>
`;
